pkg/controllers/elasticsearch/actions/scale.go

```diff
@@ -0,0 +1,212 @@
+package actions
+
+import (
+	"fmt"
+
+	apps "k8s.io/api/apps/v1beta1"
+	core "k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerror "k8s.io/apimachinery/pkg/util/errors"
+
+	"github.com/jetstack/navigator/pkg/apis/navigator/v1alpha1"
+	"github.com/jetstack/navigator/pkg/controllers"
+	"github.com/jetstack/navigator/pkg/controllers/elasticsearch/util"
+	utilapi "github.com/jetstack/navigator/pkg/util/api"
+)
+
+type Scale struct {
+	// The Elasticsearch cluster resource being operated on
+	Cluster *v1alpha1.ElasticsearchCluster
+	// The node pool being scaled
+	NodePool *v1alpha1.ElasticsearchClusterNodePool
+	// Number of replicas to scale to
+	Replicas int32
+}
+
+var _ controllers.Action = &Scale{}
+
+func (c *Scale) Name() string {
+	return "Scale"
+}
+
+func (c *Scale) Message() string {
+	return fmt.Sprintf("Scaled node pool %q to %d replicas", c.NodePool.Name, c.NodePool.Replicas)
+}
+
+// Execute will scale the s.NodePool statefulset to the desired number of
+// replicas. It will refuse to scale if the cluster is not in a 'state to scale'
+// as defined by s.canScaleNodePool.
+func (s *Scale) Execute(state *controllers.State) error {
+	if s.NodePool == nil || s.Cluster == nil {
+		return fmt.Errorf("internal error: nodepool and cluster must be set")
+	}
+
+	statefulSetName := util.NodePoolResourceName(s.Cluster, s.NodePool)
+	statefulSet, err := state.StatefulSetLister.StatefulSets(s.Cluster.Namespace).Get(statefulSetName)
+	if k8sErrors.IsNotFound(err) {
+		state.Recorder.Eventf(s.Cluster, core.EventTypeWarning, "Err"+s.Name(), "Cannot get statefulset %q for node pool %q: %v", statefulSetName, s.NodePool.Name, err)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	currentReplicas := statefulSet.Spec.Replicas
+	// TODO: not sure if we should treat nil as 1 or 0 instead of erroring
+	if currentReplicas == nil {
+		return fmt.Errorf("current number of replicas on statefulset cannot be nil")
+	}
```

Comment:
Dereferencing any nil pointer will cause a panic: https://play.golang.org/p/6eQA-QmxJWv

Reply:
Right, sorry. That was a stupid comment. But in that case, I take it that the statefulset controller doesn't handle `statefulSet.Spec.Replicas` either.

In which case I assume that there is a validation / defaulting mechanism in the API server which checks for nil / missing replicas. Ensuring that it will always have a value by the time it reaches controllers.